Fix bugs and clean up grayscale morphology operators

crisluengo opened the pull request:
<!--
Please use `pre-commit` to format code.

```
pip install pre-commit  # install the package
pre-commit install  # install git commit hook
```

Now, formatting checks will be run on each commit.
You can also run `pre-commit` manually:

```
pre-commit run -a
```
-->

## Description

This PR affects `skimage/morphology/gray.py`:
- Fixes the opening and closing to be correct (#6676).
- Makes the dilation mirror the footprint.
- Deprecates the `shift_x` and `shift_y` parameters, which don't seem meaningful after these changes.
- Makes `dilation`, `erosion`, `opening` and `closing`, with default parameters and a binary input image, identical to the equivalent `binary_...` functions. 
- Simplifies the code for `white_tophat`, and complicates the code for `black_tophat` so that it matches `white_tophat`. Both now special-case for `out is image`.
- Adds `mode` and `cval` parameters, with the special option `None` that sets the pixels outside the image to the max value for the image's dtype in the erosion, and to the min in the dilation. This causes the pixels outside the image to be ignored in these operations. This is now the default boundary condition. Previously it was `'reflect'`, because that is the default in ndimage, but that condition causes surprising results if the footprint is not simple (#6665).

This PR closes #6665 and #6676 -- it was a lot easier fixing both at the same time.

This PR causes changes to the output for some of these functions if the footprint is not mirror symmetric (especially the dilation, and it fixes the output for the opening and closing), and it affects the pixels near the image border if the footprint is not simple (i.e. not a rectangle, disk, star, etc). I would guess that 99.9% of uses of these functions involve these simple, symmetric footprints, and consequently will not notice a difference. But there surely are some people using these functions in other ways. The output will be better for these users, but it will be different. There should be a deprecation path to change the behavior, but I just don't see such a path that doesn't involve new function names, or an awkward "do the right thing" parameter. If this is totally unacceptable, you can always wait until skimage 2.0 to accept this PR.

@stefanv and @lagru: Thank you for your help with this.

## Test failure

The CI test failed on GitHub. It ran OK on my machine only if I changed the line
```python
expected = dict(np.load(fetch('data/gray_morph_output.npz')))
```
to not use `fetch`, and to have the absolute path to the file (which has changed). I think this `fetch` function is broken, it reads the file from elsewhere, not from the repository.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

crisluengo commented:
OK, I've found a way to make the desired changes without changing the output using default parameters.

The binary and gray-value version of the morphological operators are not equivalent, they can only be made equivalent by changing the output of one set.

I've updated the description above to match the new situation.

The tests still fail.

lagru commented on skimage/morphology/gray.py:
```diff
@@ -301,30 +304,32 @@ def dilation(image, footprint=None, out=None, shift_x=False, shift_y=False):
+
+    footprint = _shift_footprints(footprint, shift_x, shift_y)
+    footprint = pad_footprint(footprint, right=False)
+    if not mirror:
+        # Note that `ndi.grey_dilation` mirrors the footprint.
+        footprint = mirror_footprint(footprint)
```
```suggestion
    if not mirror:
        # Note that `ndi.grey_dilation` mirrors the footprint and this
        # additional inversion should be removed in skimage2, see gh-6676.
        footprint = mirror_footprint(footprint)
```

Correct me if I'm wrong: in #6676 we settled on not silently changing the output, so that's why we keep this "bug" from earlier? Meaning this PR shouldn't close #6676; we should keep that PR open and added to the skimage2 milestone.

crisluengo replied:
See my reply to this comment in the issue report: https://github.com/scikit-image/scikit-image/issues/6676#issuecomment-1402348378

lagru commented on skimage/morphology/gray.py:
```diff
@@ -255,20 +232,42 @@ def dilation(image, footprint=None, out=None, shift_x=False, shift_y=False):
     Returns
     -------
-    dilated : uint8 array, same shape and type as `image`
+    dilated : uint8 array, same shape and type as ``image``
```
Nitpick: 

```suggestion
    dilated : uint8 array, same shape and type as `image`
```
Currently, we have the convention to reference parameters this way. This also applies to a lot of other places in this PR.

lagru commented on skimage/morphology/gray.py:
```diff
@@ -176,19 +115,41 @@ def erosion(image, footprint=None, out=None, shift_x=False, shift_y=False):
     -----
     For ``uint8`` (and ``uint16`` up to a certain bit-depth) data, the
-    lower algorithm complexity makes the `skimage.filters.rank.minimum`
+    lower algorithm complexity makes the ``skimage.filters.rank.minimum``
```
```suggestion
    lower algorithm complexity makes the :func:`skimage.filters.rank.minimum`
```
Good opportunity to create a linked reference here (and elsewhere through this PR) if we are already changing it up.

lagru reviewed (commented):
Thanks @crisluengo for this! I only started reviewing this but you already got a lot right by yourself! :ok_hand: Please feel welcome to ask for clarification or discuss if you don't agree with my suggestions.

crisluengo commented on skimage/morphology/gray.py:
```diff
@@ -222,19 +186,32 @@ def erosion(image, footprint=None, out=None, shift_x=False, shift_y=False):
     if out is None:
         out = np.empty_like(image)
 
+    if mode is None:
+        mode = "constant"
+        if np.issubdtype(out.dtype, bool):
+            cval = True
+        elif np.issubdtype(out.dtype, np.integer):
+            cval = np.iinfo(image.dtype).max
+        else:
+            cval = np.inf
```
`ndi.grey_erosion` doesn't have this boundary extension mode. You could call it "max" and "min", but it's nice to have one boundary extension mode that means "max" in the erosion and "min" in the dilation. Because now we can use this mode in the opening and closing, and get the right behavior there. So I called it `None`. I would make this the default mode in skimage2, because it's the mode that makes most sense for all morphological operators, it never causes weird artefacts at the image boundary.

Of course we could suggest adding this mode to `ndimage`...

lagru replied:
Ah, that sound's useful indeed. Though I would prefer to be a bit more specific about the name. That way we can also preserve `None` in case we ever need it for a deprecation or something else. What do you think about the naming the modes "dtype_max" and "dtype_min" respectively?

I think it wouldn't hurt to ask SciPy if they are interested to add this mode. If they are interested we should probably agree on a common name for this mode. But waiting for SciPy shouldn't keep us from adding it in this PR in the meantime.

crisluengo replied:
I've added "max" and "min" modes, and an "ignore", which is interpreted as "max" by the erosion and as "min" by the dilation. This makes it possible to pass "ignore" to the opening or closing, and get the right behavior.

lagru commented on skimage/morphology/gray.py:
```diff
@@ -1,39 +1,40 @@
-def _iterate_gray_func(gray_func, image, footprints, out):
-    """Helper to call `binary_func` for each footprint in a sequence.
+def _iterate_gray_func(gray_func, image, footprints, out, mode, cval):
+    """Helper to call ``binary_func`` for each footprint in a sequence.
```
```suggestion
    """Helper to call `gray_func` for each footprint in a sequence.
```

lagru commented on skimage/morphology/gray.py:
```diff
@@ -1,39 +1,40 @@
-    binary_func is a binary morphology function that accepts "structure",
-    "output" and "iterations" keyword arguments
-    (e.g. `scipy.ndimage.binary_erosion`).
+    binary_func is a binary morphology function that accepts "footprint",
```
```suggestion
    `gray_func` is a binary morphology function that accepts "footprint",
```

lagru reviewed (commented):
I'm currently thinking about how can keep our API correct and consistent in the long run. How do we justify adding the the `mirror` parameter only to some of our functions that use a `footprint` whose symmetry would impact their output. Rather than let a bug influence our API I would prefer to skip the perhaps a bit trivial `mirror` parameter and enable the user to do so beforehand with `mirror_footprint`.

Maybe we could add `mirror` temporarily to `binary_dilation` as part of a deprecation path? E.g.

- Release x: In `binary_dilation`, add `mirror=True`. Using `True` will raise a `FutureWarning` that mirroring will be removed in the future. Advise the user to use `mirror=False` and apply `mirror_footprint` if they need it.
- Release x+1: Remove the mirroring behavior. Using the `mirror` parameter at all raises a `FutureWarning` that the parameter is deprecated. 
- Release x+2: Remove the `mirror` parameter.

For users that use `footprint=None` anyway, we could reduce the noise by not raising deprecation warnings at all in that case. What do you think?

crisluengo commented:
> I'm currently thinking about how to can keep our API correct and consistent in the long run. How do we justify adding the the `mirror` parameter only to some of our functions that use a `footprint` whose symmetry would impact their output. Rather than let a bug influence our API I would prefer to skip the perhaps a bit trivial `mirror` parameter and enable the user to do so beforehand with `mirror_footprint`.

Yes, this would be cleaner


> Maybe we could add `mirror` temporarily to `binary_dilation` as part of a deprecation path? E.g.
> 
> * Release x: In `binary_dilation`, add `mirror=True`. Using `True` will raise a `FutureWarning` that mirroring will be removed in the future. Advise the user to use `mirror=False` and apply `mirror_footprint` if they need it.
> * Release x+1: Remove the mirroring behavior. Using the `mirror` parameter at all raises a `FutureWarning` that the parameter is deprecated.
> * Release x+2: Remove the `mirror` parameter.

I'm not sure about forcing people to add a `mirror` argument to their function call to silence a warning, and then make that parameter produce a warning the next release.

I am not sure there is a clean way to change this behavior other than changing the function name (i.e. adding a new function, deprecating the old one). I think the simplest solution is to keep the behavior as it is now, and document the difference between `binary_dilation` and `dilation` (I've already added this to the docs as a note). For skimage 2 we could change the behavior.

crisluengo commented on skimage/morphology/binary.py:
```diff
@@ -52,6 +55,18 @@ def binary_erosion(image, footprint=None, out=None):
+        .. versionadded:: 0.21
+            `mirror` was added in 0.21.
+
+    border_value : bool or None, optional
+        Value at the border in the output array. ``None`` is the same as
+        ``True``. Default is ``True``.
```
Doh! I copied the ndimage documentation for the parameter without reading it critically. You are right, this is not at all good. 🤦

crisluengo commented on skimage/morphology/binary.py:
```diff
@@ -169,6 +223,13 @@ def binary_opening(image, footprint=None, out=None):
     out : ndarray of bool, optional
         The array to store the result of the morphology. If None
         is passed, a new array will be allocated.
+    border_value : bool or None, optional
+        Value at the border in the output array. ``None`` is ``True`` for the
+        erosion and ``False`` for the dilation, causing pixels outside the
+        image not to influence the results. Default is ``None``.
```
This would allow setting `[True, True]`, `[False, False]`, `[True, False]` and `[False, True]`. Three of those options correspond to `True`, `False` and `None` in the current iteration, the fourth option doesn't seem especially interesting to me. It makes the parameter more complex, I'm not sure the added option is worth that complexity.

What we could do however is change this for a `mode` parameter like in the gray-scale case, with options "max", "min" and "ignore".

lagru replied:
Sounds good and seems more consistent to to have a `mode` parameter everywhere. :+1:

lagru commented:
> I think the simplest solution is to keep the behavior as it is now, and document the difference between binary_dilation and dilation (I've already added this to the docs as a note). For skimage 2 we could change the behavior.

I can live with that. :sweat_smile: 

@scikit-image/core do we already collect these TODO's for skimage2 somewhere in a more structured and explicit list than our milestone? Otherwise we should probably start such a document sooner rather than later..

mkcor commented:
> do we already collect these TODO's for skimage2 somewhere in a more structured and explicit list than our milestone?

@lagru I started this [HackMD](https://hackmd.io/@iJzCdxsaRj2x_eVWgqy98Q/HyYOmLnaO) for this very purpose! See links therein.

crisluengo commented:
@stefanv Thanks for helping with this PR.

I have addressed the issues that @lagru raised:
- Both binary and gray-scale morphology functions now have a `mode` parameter to determine what happens outside the image boundary.
- There no longer is a `mirror` parameter, the user can explicitly mirror the footprint using the new mirroring function.

The functions should only produce different values they did previously in the case of even-sized footprints, where previously the results were wrong (so differences are because we're fixing a bug) -- if you want there to be no changes, then there's no point in fixing bugs.

crisluengo commented:
By the way: tests pass locally only when I change the test code to load the updated `skimage/data/gray_morph_output.npz` out of the repository. Without the change, I think it fetches it elsewhere and it's not using the updated file. Again, the updated file is necessary because the output for openings and closings with even-sized footprints changes.

lagru commented:
@crisluengo, sorry for dropping the ball on this for a bit. Should I go ahead and address the merge conflicts? This looks close to ready to me and hope that we can finish this up in the next few days. :crossed_fingers:

crisluengo commented:
@lagru Sorry for dropping the ball on this one. I've merged in main, and updated mentions of 0.22 to 0.23, so this should be ready to include in the next release.

lagru commented on skimage/morphology/gray.py:
```diff
@@ -70,93 +71,48 @@ def _shift_footprint(footprint, shift_x, shift_y):
+    return _shift_footprint(footprint, shift_x, shift_y)
+
+
+def _handle_mode_and_cval(dtype, mode, cval):
```
How about the name `_min_max_to_constant_mode` instead? It's less generic than "handle".

lagru commented on skimage/morphology/gray.py:
```diff
@@ -70,93 +71,48 @@ def _shift_footprint(footprint, shift_x, shift_y):
+
+
+def _handle_mode_and_cval(dtype, mode, cval):
+    """Sets the mode and cval parameters for 'max' and 'min'."""
```
```suggestion
    """Replace 'max' and 'min' with appropriate 'cval' and 'constant' mode."""
```

lagru commented on skimage/morphology/gray.py:
```diff
@@ -70,93 +71,48 @@ def _shift_footprint(footprint, shift_x, shift_y):
-        if padding:
-            image = np.pad(image, pad_widths, mode='edge')
-            out_temp = np.empty_like(image)
+    if _footprint_is_sequence(footprint):
+        return tuple((_shift_footprint(fp, shift_x, shift_y), n)
+                     for fp, n in footprint)
```
Might want to use recursion here is well. Though, it's less clear if it's more readable. :shrug:

crisluengo replied:
`_shift_footprints` has a bit more code than in the other cases. Combining these would repeat the check for `shift_x` and `shift_y`.

Not sure if the warning would be printed repeatedly, does Python only print something the first time?

lagru commented on skimage/morphology/gray.py:
```diff
@@ -175,20 +131,35 @@ def erosion(image, footprint=None, out=None, shift_x=False, shift_y=False):
     out : ndarrays, optional
         The array to store the result of the morphology. If None is
         passed, a new array will be allocated.
-    shift_x, shift_y : bool, optional
-        shift footprint about center point. This only affects
-        eccentric footprints (i.e. footprint with even numbered
-        sides).
```
Sorry, it's been a while and I may have forgotten previous discussions. According to the PR descriptions, shouldn't `erosion` and other function have a new `mirror` parameter to replace this one?

crisluengo replied:
Somewhere we decided it was better to let the user call `mirror_footprint` themselves. It keeps things simpler.

lagru reviewed (commented):
@crisluengo, thanks for picking it up again! I am starting to get back into this. This isn't the most straightforward PR to review, so apologies if the review is also all over the place. :sweat_smile:

Considering that it's been only me doing the reviews so far, if you see an opportunity to separate this into multiple PRs please go for it. That might help a lot with getting this through a lot faster.

lagru commented:
I think the CI failure is due to a recent update to Cython.

crisluengo commented on skimage/data/gray_morph_output.npz:
I updated the comment referencing the PR.

crisluengo commented on skimage/data/_registry.py:
```diff
@@ -133,7 +133,7 @@
     "data/foo3x5x4indexed.png": "48a64c25c6da000ffdb5fcc34ebafe9ba3b1c9b61d7984ea7ca6dc54f9312dfa",
     "data/mssim_matlab_output.npz": "cc11a14bfa040c75b02db32282439f2e2e3e96779196c171498afaa70528ed7a",
-    "data/gray_morph_output.npz": "3012eb994e864e1dca1f66fada6b4375f84eac63658d049886b710488c2394d1",
+    "data/gray_morph_output.npz": "49a0dae607cd8d31e134b4bfcbf0d86b13751fdce9667d8bf1ade93d435191b1",
```
The output changes for openings and closings with even-sized kernels, where previously the result was wrong (not a closing or opening). This is the main reason for the PR.

It also changes for other asymmetric kernels, but there are no tests for those.

lagru replied:
Okay, so to state the obvious for future readers, we consider it fine then to change the output, because we consider the previous behavior a bug.

crisluengo commented:
@lagru 
> Considering that it's been only me doing the reviews so far, if you see an opportunity to separate this into multiple PRs please go for it. That might help a lot with getting this through a lot faster.

Considering I thought this was going to be maybe a 1 hour effort when I started, and I've now put it at least 10, I think I'm not going to agree to a large effort like this, sorry. I know it's hard to review PRs.

crisluengo commented:
I'm seeing the same build errors when trying to build skimage locally. I looked at the C source code generated by Cython, it calls a function over and over again that is not declared, and defined at the bottom of the same file. I don't think this has ever been correct C, the implicit declaration would have assumed an `int` return type, which is not what the function returns.

But maybe it sort of worked before C99? (I do see "error: implicit declaration of function '__pyx_fuse_0__pyx_f_7skimage_7_shared_13interpolation_get_pixel2d' is invalid in C99"). Was the GCC compiler recently updated to default to C99, or to be more strict with these implicit declarations?

If I downgrade Cython I see the exact same errors. I went all the way back to 0.29.26 (Dec 16, 2021), but it didn't fix the issue.

crisluengo commented:
`fetch('data/gray_morph_output.npz')` seems to be getting a different file than the one in  this PR, as the hash doesn't match the newly updated one in `data/_registry.py`. I don't know where `fetch` downloads stuff from, but it seems silly not to use the locally available copy.

Also, `skimage/measure/tests/test_fit.py::test_ellipse_parameter_stability` is now failing. I haven't touched that test, and I don't think it uses any morphology functionality, so I don't think it's caused by this PR. I can get the test to pass by changing the failing assertion at line 247 into:
```python
assert_almost_equal(theta_prime % (np.pi/2), theta % (np.pi/2))
```
(adding the modulo half pi, which seems somehow wrong, it means the long and short axes are switched).

lagru commented on skimage/morphology/footprints.py:
```diff
@@ -959,3 +959,69 @@ def star(a, dtype=np.uint8):
+    return footprint[(slice(None, None, -1),) * footprint.ndim]
+
+
+def pad_footprint(footprint, *, pad_end=True):
```
I refactored the parameter `right` to `pad_end` and also made it keyword only. Basically see the description in https://github.com/scikit-image/scikit-image/pull/6695/commits/1bc9a7ea2588472475cd2a937996caf7d737cdb5 for the reasoning. Just highlighting this because it doesn't hurt to deliberate a bit about new API we are introducing.

crisluengo replied:
Yes, this is much better!

lagru reviewed (commented):
Quick report on the review progress: I am checking the PR locally so I took the liberty to apply some changes directly between https://github.com/scikit-image/scikit-image/pull/6695/commits/1bc9a7ea2588472475cd2a937996caf7d737cdb5 to https://github.com/scikit-image/scikit-image/pull/6695/commits/8ae5862f76eccf58d956776fafb3dea93d36dde1 since it seemed less effort for us both in total and easier than explaining. Please feel free to revert, discuss or further tweak. 

I'm probably done with `pad_footprint` and `mirror_footprint` now. Getting there for the rest. :wink: 

And let me know in case you feel more comfortable with making changes yourself upon feedback.

crisluengo commented on skimage/morphology/binary.py:
```diff
@@ -79,16 +99,24 @@ def binary_erosion(image, footprint=None, out=None):
+            ndi.binary_erosion, image, footprint, out, border_value
+        )
+
+    print(f"binary_erosion, {border_value=}")
```
Good catch! Ouch!

crisluengo commented on skimage/data/gray_morph_output.npz:
Yes, you are right about the examples in #6676. Here's what's happening:

The default boundary extension, `mode="reflect"`, is applied for the dilation, then it's applied again, separately, for the erosion. This makes it possible for some unexpected results when the SE reaches outside the image. The ideal way to apply such a boundary extension is to extend the image, apply the dilation (which now also updates pixel values in the extended boundary), then apply the erosion, then cut off the boundary region again.

When using the new `mode="ignore"`, those tests do pass. This would be the correct default boundary extension, but we decided we didn't want to change how the filters were working outside of fixing the non-symmetric kernel case. For the binary version of the operators, the default boundary extension indeed is `mode="ignore"`.

Updated test case:
```python
import skimage
import numpy as np

# A test image
img = np.random.rand(300, 300)
img = skimage.filters.gaussian(img, 5)
img = (img > 0.5).astype(np.uint8)
img *= 255

# A non-symmetric SE
se = np.array([[0,0,1],[0,1,1],[1,1,1]])

# Closing must be extensive, but it isn't
closing = skimage.morphology.closing(img, se, mode="ignore")
assert(np.all(closing >= img))

# Opening must be anti-extensive, but it isn't
opening = skimage.morphology.opening(img, se, mode="ignore")
assert(np.all(img >= opening))

# Again for the binary case
img = img > 0

# Closing must be extensive, it is
closing = skimage.morphology.binary_closing(img, se)
assert(np.all(closing >= img))

# Opening must be anti-extensive, it is
opening = skimage.morphology.binary_opening(img, se)
assert(np.all(img >= opening))
```

lagru replied:
Ah, I think I get it now. Thanks for the patience and explanation! :) Your updated snippet was a perfect candidate to turn into additional tests (see https://github.com/scikit-image/scikit-image/pull/6695/commits/768a301c130e1799078355c11db70ce98aa876af).

lagru commented:
I also added an input guard for the `mode` parameter in https://github.com/scikit-image/scikit-image/pull/6695/commits/9bab0e46cbbf1755b2b1006f1c3e4e742b22eb70. I noticed that without it e.g. `binary_closing(..., mode="reflect")` didn't throw any error; it was internally as "ignore". We should be explicit about it.

lagru commented on skimage/morphology/footprints.py:
```diff
@@ -962,3 +962,69 @@ def star(a, dtype=np.uint8):
+    Parameters
+    ----------
+    footprint : ndarray or tuple
+        The input footprint or sequence of footprints
```
```suggestion
        A footprint expressed as a 2-D array of 1's and 0's. The footprint can
        also be provided as a sequence of smaller footprints.
```

lagru commented on skimage/morphology/footprints.py:
```diff
@@ -962,3 +962,69 @@ def star(a, dtype=np.uint8):
+    Parameters
+    ----------
+    footprint : ndarray or tuple
+        The input footprint or sequence of footprints
```
```suggestion
        A footprint expressed as a 2-D array of 1's and 0's. The footprint can
        also be provided as a sequence of smaller footprints.
```

lagru commented on skimage/morphology/gray.py:
```diff
@@ -199,8 +191,12 @@ def erosion(image, footprint=None, out=None, shift_x=False, shift_y=False):
+    a footprint sequence of this type.
+
+    For even-sized footprints, :func:`skimage.morphology.binary_erosion` and
+    this function produce an output that differs: one is shifted by one pixel
```
```suggestion
    this function produces an output that differs: one is shifted by one pixel
```

lagru commented on skimage/data/_registry.py:
```diff
@@ -133,7 +133,7 @@
     "data/foo3x5x4indexed.png": "48a64c25c6da000ffdb5fcc34ebafe9ba3b1c9b61d7984ea7ca6dc54f9312dfa",
     "data/mssim_matlab_output.npz": "cc11a14bfa040c75b02db32282439f2e2e3e96779196c171498afaa70528ed7a",
-    "data/gray_morph_output.npz": "3012eb994e864e1dca1f66fada6b4375f84eac63658d049886b710488c2394d1",
+    "data/gray_morph_output.npz": "49a0dae607cd8d31e134b4bfcbf0d86b13751fdce9667d8bf1ade93d435191b1",
```
Done. See https://gitlab.com/scikit-image/data/-/merge_requests/23.

lagru commented on skimage/morphology/gray.py:
```diff
@@ -1,33 +1,33 @@
     """
     fp, num_iter = footprints[0]
-    gray_func(image, footprint=fp, output=out)
+    gray_func(image, footprint=fp, output=out, mode=mode, cval=cval)
```
See https://github.com/scikit-image/scikit-image/pull/6695/commits/4a31f39bf984952c3cfbb4ee811248bfb8777ecb and https://github.com/scikit-image/scikit-image/pull/6695/commits/eb02ed77cb4e5c2ca824b17980afc2396ddde08f.

lagru reviewed (approved):
Approving on the condition that https://gitlab.com/scikit-image/data/-/merge_requests/23 is merged. :tada: 

@crisluengo, I took the liberty to tweak / add a bit which I think was the way faster option for this PR. Please feel welcome to review or revert. Thank you very much for your work and especially clarifications. :D

crisluengo commented:
@lagru Thank you so much for your work on this PR.

Next time I'll make sure to keep my PRs small. :)

stefanv commented on skimage/morphology/binary.py:
```diff
@@ -110,6 +140,15 @@ def binary_dilation(image, footprint=None, out=None):
     out : ndarray of bool, optional
         The array to store the result of the morphology. If None is
         passed, a new array will be allocated.
+    mode : str, optional
```
This argument name caught me by surprise, since usually `mode` defines how boundary values are obtained: `{'reflect', 'constant', 'edge', 'symmetric', 'wrap'}`.

crisluengo replied:
It is exactly the same meaning, but the options are different. The ndimage function we call here only has the option of a constant value (you can either pad with 0s or with 1s). We’re using the strings ‘min’ and ‘max’ for those two options, they match the strings used in the gray-scale morphology. And we added the ‘ignore’ option (which we also added to the gray-scale version), which is ‘max’ for erosion and ‘min’ for dilation.

We cannot easily implement ‘reflect’ and the other more complex modes.  Though these are available in the gray-scale versions of the operators, which we could be using here instead. They all do the same thing for binary images anyway.

stefanv reviewed (commented):
Overall, the PR looks good to me, although I suspect there are a lot of subtleties.

@crisluengo Any chance you could join us tomorrow morning 9am Pacific for the community call to discuss this PR?

https://scientific-python.org/calendars/

crisluengo commented:
@stefanv I might be able to hop on that call for a bit, if you think it’d be useful.

stefanv commented:
> @stefanv I might be able to hop on that call for a bit, if you think it’d be useful.

I'll be around from 9:15; a quick "this is the problem, here's all the things I did to fix it, here's what remains to be done in the future" would be great.

We need to make sure all those future TODOs are documented, either in an issue or in the TODO.txt.

crisluengo commented:
@stefanv OK, I’ll hop on the call 9:15 Pacific.

stefanv commented:
Thank you Cris! And thanks, Lars, for carefully reviewing and shepherding the PR through.

lagru commented:
Now we need to resolve https://github.com/scientific-python/changelist/issues/45 so we can represent this properly in our release notes. Otherwise, we have to do so manually (see updated PR description).
